### PR TITLE
Make the functions which produce the components of the sighash reusable.

### PR DESCRIPTION
**Pull request by nuttycom**
When implementing transaction nonmalleability, it will be useful
to have this already done.

### CODE REVIEW

**Review by daira — APPROVED** [review]
utACK

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ jobs:
 
       - name: Fetch path to Zcash parameters
         working-directory: ./zcash_proofs
```
Comment (str4d) [inline .github/workflows/ci.yml:21]: I think this is needed to get Windows working:
```suggestion
        working-directory: ./zcash_proofs
        shell: bash
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 2]
```diff
@@ -77,9 +79,9 @@ impl SigHashVersion {
     }
 }
 
-fn prevout_hash(tx: &TransactionData) -> Blake2bHash {
-    let mut data = Vec::with_capacity(tx.vin.len() * 36);
-    for t_in in &tx.vin {
+fn prevout_hash(vin: &Vec<TxIn>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:82]: Always pass a slice in place of an immutable reference to a `Vec`:
```suggestion
fn prevout_hash(vin: &[TxIn]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 3]
```diff
@@ -88,9 +90,9 @@ fn prevout_hash(tx: &TransactionData) -> Blake2bHash {
         .hash(&data)
 }
 
-fn sequence_hash(tx: &TransactionData) -> Blake2bHash {
-    let mut data = Vec::with_capacity(tx.vin.len() * 4);
-    for t_in in &tx.vin {
+fn sequence_hash(vin: &Vec<TxIn>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:93]:
```suggestion
fn sequence_hash(vin: &[TxIn]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 4]
```diff
@@ -101,9 +103,9 @@ fn sequence_hash(tx: &TransactionData) -> Blake2bHash {
         .hash(&data)
 }
 
-fn outputs_hash(tx: &TransactionData) -> Blake2bHash {
-    let mut data = Vec::with_capacity(tx.vout.len() * (4 + 1));
-    for t_out in &tx.vout {
+fn outputs_hash(vout: &Vec<TxOut>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:106]:
```suggestion
fn outputs_hash(vout: &[TxOut]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 5]
```diff
@@ -121,28 +123,32 @@ fn single_output_hash(tx_out: &TxOut) -> Blake2bHash {
         .hash(&data)
 }
 
-fn joinsplits_hash(tx: &TransactionData) -> Blake2bHash {
+fn joinsplits_hash(
+    txversion: u32,
+    joinsplits: &Vec<JSDescription>,
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:128]:
```suggestion
    joinsplits: &[JSDescription],
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 6]
```diff
@@ -121,28 +123,32 @@ fn single_output_hash(tx_out: &TxOut) -> Blake2bHash {
...
     Blake2bParams::new()
         .hash_length(32)
         .personal(ZCASH_JOINSPLITS_HASH_PERSONALIZATION)
         .hash(&data)
 }
 
-fn shielded_spends_hash(tx: &TransactionData) -> Blake2bHash {
-    let mut data = Vec::with_capacity(tx.shielded_spends.len() * 384);
-    for s_spend in &tx.shielded_spends {
+fn shielded_spends_hash(shielded_spends: &Vec<SpendDescription>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:149]:
```suggestion
fn shielded_spends_hash(shielded_spends: &[SpendDescription]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 7]
```diff
@@ -155,9 +161,9 @@ fn shielded_spends_hash(tx: &TransactionData) -> Blake2bHash {
         .hash(&data)
 }
 
-fn shielded_outputs_hash(tx: &TransactionData) -> Blake2bHash {
-    let mut data = Vec::with_capacity(tx.shielded_outputs.len() * 948);
-    for s_out in &tx.shielded_outputs {
+fn shielded_outputs_hash(shielded_outputs: &Vec<OutputDescription>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:164]:
```suggestion
fn shielded_outputs_hash(shielded_outputs: &[OutputDescription]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 8]
```diff
@@ -166,9 +172,9 @@ fn shielded_outputs_hash(tx: &TransactionData) -> Blake2bHash {
         .hash(&data)
 }
 
-fn tze_inputs_hash(tx: &TransactionData) -> Blake2bHash {
+fn tze_inputs_hash(tze_inputs: &Vec<TzeIn>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:175]:
```suggestion
fn tze_inputs_hash(tze_inputs: &[TzeIn]) -> Blake2bHash {
```

File: zcash_primitives/src/transaction/sighash.rs [hunk 9]
```diff
@@ -177,9 +183,9 @@ fn tze_inputs_hash(tx: &TransactionData) -> Blake2bHash {
         .hash(&data)
 }
 
-fn tze_outputs_hash(tx: &TransactionData) -> Blake2bHash {
+fn tze_outputs_hash(tze_outputs: &Vec<TzeOut>) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:186]:
```suggestion
fn tze_outputs_hash(tze_outputs: &[TzeOut]) -> Blake2bHash {
```

**Review by str4d — CHANGES_REQUESTED** [review]

File: zcash_primitives/src/transaction/sighash.rs (hunk 2 above)
Comment (nuttycom) [inline zcash_primitives/src/transaction/sighash.rs]:
```suggestion
fn prevout_hash(vin: &[TxIn]) -> Blake2bHash {
```

**Review by str4d — APPROVED** [review]

**Pull request merged**